### ISSUE TITLE
[question] cursor performance

### ISSUE THREAD

**Issue by CristianPi** [Venemo/node-lmdb#170]
Hi,
i was making some benchmarks (32GB ram, 1TB nvme) about this project,
i see a large difference between cursors and gets (Binary keys/values).

like, 1M read/sec for normal random gets
and about 0.5M for cursors reads

Shouldn't cursor performance be the same or greater than normal reads?
Is this related to the c++ <-> js overhead? 

example code:

`function query(store: string, field: string, value: any): any[] {

    let cursor = new lmdb.Cursor(readtx, dbi, { keyIsBuffer: true });

    const start = store + '_' + field + '_' + value;

    const stop = store + '_' + field + '_' + (value+1);

    const stopBuffer = Buffer.from(stop);

    console.log(start,stop);

    let list:any = []

    for (let found = cursor.goToRange(Buffer.from(start)); found !== null; found = cursor.goToNext()) {
        const key = cursor.getCurrentBinary();
        list.push(key);
    }

    return list;
}`

the function is called only once on a ~10M keys db.

**Comment by CristianPi** [comment]
you can find the example code here:
[https://github.com/CristianPi/benchlmdb.git](url)
i extended **node-lmdb/benchmark/index.js**

for convenience the results on my machine:
```
getBinary x 1,027,705 ops/sec ±2.07% (87 runs sampled)
getBinaryUnsafe x 659,704 ops/sec ±12.10% (75 runs sampled)
getString: 
getStringUnsafe x 2,134,603 ops/sec ±7.96% (91 runs sampled)
cursorGoToNext x 819,103 ops/sec ±7.21% (45 runs sampled)
cursorGoToNextgetCurrentBinary x 433,634 ops/sec ±5.03% (66 runs sampled)
Fastest is getStringUnsafe
```

**Comment by kriszyp** [comment]
Thank you for putting together this benchmark. I took a look at these benchmarks to see what might be the bottlenecks. From what I can tell, it is actually less about C++/JS overhead and more about the overhead of creating Buffer objects. And by default (and when you have keyIsBuffer) cursor.goToNext() returns the new current key as a buffer. This and the getCurrentBinary both end up creating buffer objects. By switching to keyIsString for the cursor, which causes goToNext (and all the cursor methods) to return a string and using getCurrentString, I got about 6Mops/s and 3Mop/s, which is indeed faster than random access gets.

This is certainly interesting and counter-intuitive to me. I would have thought that avoiding the construction of a string would mean buffers are faster, but that certainly does not seem to be the case.

It is possible that further optimization could be achieved by doing some type of batched cursor operation that returns a set of values, but the most significant direct bottleneck here seems to be Buffer object creation.

**Comment by Venemo** [comment]
What happens if the keys are integers? Does that improve performance further, compared to strings?

**Comment by kriszyp** [comment]
Yes, I get about a 20% speed up from using the keyIsUint32 keys instead of strings.

**Comment by CristianPi** [comment]
it's a big limitation of nodejs.
It **should** be possible to use "string" as buffers (to avoid the overhead and do sorting on fixed size intergers with prefix ect. ect.), practically solving the issue, without increasing the storage size.
On server scale even a 5% improvement is important, here we are talking 100% / 600%.
So i will try to go with this route,
but it's either  more storage size, or complicated logic on the application layer (that probably will have performance implication) for "transforming" buffers <-> strings.


> Yes, I get about a 20% speed up from using the keyIsUint32 keys instead of strings.

the only limitation is the 4 byte space, but, still, it does not seem to be that big improvement, if i want to use 8 byte space (a.k.a 2 Uint32) the performance should drop.




> It is possible that further optimization could be achieved by doing some type of batched cursor operation that returns a set of values, but the most significant direct bottleneck here seems to be Buffer object creation.

It should be possible making prefetch requests in a different thread, soaking the buffer creation cost, but is it worth it? and it will be only useful for a single thread application, having multiples process reading from the DB will make the performance gain negative or zero approaching the 100% cpu usage.
Another alternative is having a pools of buffers and then reading the values directly from lmdb inside those buffers (just a memcpy). the problem will be how many buffers and of what size (specified by the application layer or increased automatically), then the actual size of the readed value should be solved on the application layer (or returned as an extra parameter).
This solution is very use case specific, and have a tons of drawback (it will be like reading with getBinaryUnsafe, fixed memory consumption), but still, should be faster.
I will try to make some benchmark.

Strings seems the only easy solution for performance oriented projects (beside patching Node or prefetch requests).
I will try this too.

Thanks for the time invested in this issue :1st_place_medal:

**Comment by kriszyp** [comment]
So I actually spent some time prototyping a function (in native/C) that could retrieve a range of entries using a start and end key. I thought this might provide a big performance improvement by reducing the number of JS calls to the native functions. I ran it with benchmarks, retrieving 100 entries at once (and putting them in an array to return). It ended being rather disappointing, it really only improved performance of per-entry retrieval by about 3-4% versus calling goToNext + getCurrentString for every entry. I guess the morale of the story is that the calling native code is pretty fast, it is the creation of the V8 values/objects that takes most of the time (beyond the LMDB execution itself).

>It should be possible to use "string" as buffers ...  do sorting on fixed size intergers 

So there are a couple other related things I have been looking into. First, I think node-lmdb only supports UTF-16 natively. There is a slight (tiny) performance advantage to UTF-16 with isolated benchmarks, but in the broader picture of data storage size, and performance implications of page and cache faults that are dependent on data size, I believe UTF-8 is almost certainly a much better choice encoding choice (that is what we use, but we have also done the conversion from buffers in JS, and I see now that is definitely has a performance cost).
Another package we built/use is ordered-binary which converts JS primitives to buffers with type preservation and consistent binary ordering so you get identical db sorting as JS comparisons (for numbers and strings). Currently it is written in JS, but I was looking into making that work natively in C so that you could use this a key conversion without going through Node buffers and get automatic conversion from JS primitives with consistent ordering. However, that is a little higher level design, so will probably go in the lmdb-store that builds on top node-lmdb, rather than node-lmdb itself.

> It should be possible making prefetch requests in a different thread, soaking the buffer creation cost, but is it worth it?

Right. There could be some value in doing database interaction on a separate thread, but only if there is a sufficient amount of work that it isn't more work than the event queuing to notify the main thread of work completed. Because LMDB is so fast, you would probably need to retrieve over 100 entries at once before you really benefited from a separate thread (using a separate thread for commit writes, on the other hand, is a big win). And FWIW, allocating the memory blocks for the node buffers is cheap, it seems it is the construction of Node buffer wrapper objects that is more expensive, and I believe that operation has to be done on the main thread since it involves interaction with V8 objects.

> Another alternative is having a pools of buffers and then reading the values directly from lmdb inside those buffers (just a memcpy).

Indeed that would probably be quite fast, and is a clever idea. But like you said, depending on the application, and how dynamic the data size is, it could be terribly complex and with a lot of application specific code that we seems undesirable for most general library usage.
